src/mp4.zig

```diff
@@ -0,0 +1,212 @@
+    /// the atom size (including the header size of 8 bytes)
+    size: u32,
+    /// the name or type
+    name: [4]u8,
+};
+
+fn readAtomHeader(reader: anytype, seekable_stream: anytype) !AtomHeader {
+    const size = switch (try reader.readIntBig(u32)) {
+        0 => blk: {
+            // a size of 0 means the atom extends to end of file
+
+            const remaining = (try seekable_stream.getEndPos()) - (try seekable_stream.getPos());
+
+            break :blk @intCast(u32, remaining);
+        },
+        1 => {
+            // a size of 1 means the atom header has an extended size field
+            // TODO: implement this if relevant ?
+
+            std.debug.panic("unhandled 64 bit size", .{});
+        },
+        else => |n| n,
+    };
+
+    var name: [4]u8 = undefined;
+    _ = try reader.readAll(&name);
+
+    return AtomHeader{
+        .size = size,
+        .name = name,
+    };
+}
+
+/// Generic data atom
+///
+/// Its value should be interpreted based on the type indicator and locale indicator.
+/// For our purposes it's probably enough to treat them as UTF-8 strings.
+const DataAtom = struct {
+    type_indicator: u32,
+    locale_indicator: u32,
+    value: []u8,
+};
+
+fn readDataAtom(allocator: Allocator, reader: anytype, seekable_stream: anytype) !DataAtom {
+    const atom_header = try readAtomHeader(reader, seekable_stream);
+    if (atom_header.size < 8) return error.InvalidDataAtom;
+    if (!std.mem.eql(u8, "data", &atom_header.name)) return error.InvalidDataAtom;
+
+    // -8 for the atom header size
+    // -8 for the data atom "metadata" (type indicator and locale indicator)
+    const data_size = atom_header.size - 8 - 8;
+
+    var res = DataAtom{
+        .type_indicator = try reader.readIntBig(u32),
+        .locale_indicator = try reader.readIntBig(u32),
+        .value = try allocator.alloc(u8, data_size),
+    };
+    errdefer allocator.free(res.value);
+
+    const n = try reader.readAll(res.value);
+    if (n < data_size) return error.EndOfStream;
+
+    return res;
+}
+
+const MetadataAtom = struct {
+    name: []const u8,
+    export_name: []const u8 = "",
+};
+
+// zig fmt: off
+const metadata_atoms = &[_]MetadataAtom{
+    .{ .name = "\xA9nam", .export_name = "track"        },
+    .{ .name = "\xA9alb", .export_name = "album"        },
+    .{ .name = "\xA9ART", .export_name = "artist"       },
+    .{ .name = "aART"   , .export_name = "album_artist" },
+    .{ .name = "\xA9des", .export_name = "description"  },
+    .{ .name = "\xA9day", .export_name = "release_date" },
+    .{ .name = "\xA9cmt", .export_name = "comment"      },
+    .{ .name = "\xA9too", .export_name = "tool"         },
+    .{ .name = "\xA9gen", .export_name = "genre"        },
+    .{ .name = "\xA9wrt", .export_name = "composer"     },
+    .{ .name = "\xA9cpy", .export_name = "copyright"    },
+};
+// zig fmt: on
+
+fn getMetadataAtom(name: []const u8) ?MetadataAtom {
+    for (metadata_atoms) |atom| {
```

Comment:
This can be an `inline for` I believe (alternatively, `metadata_atoms` could use `std.ComptimeStringMap` but given that all the names are the same length, it won't have much benefit over `inline for`).

Reply:
Yep, done